aten/src/ATen/native/mps/operations/Indexing.mm

```diff
@@ -28,6 +28,152 @@
 namespace at {
 namespace native {
 
+static
+void getDataOffsets(const TensorIteratorBase& iter, simd_uint3* data_offsets, uint32_t num_outputs) {
+  // Collect offsets for Input, Output and Index tensors
+  const uint32_t nOffsets = 3;
+  const uint32_t nDim = iter.ndim();
+
+  std::vector<uint32_t> strides[nDim];
+  const IntArrayRef& iterShape = iter.shape();
+
+  for (uint32_t i=0; i < nDim; i++)
+    for (uint32_t offset = 0; offset < nOffsets; offset++)
+      strides[i].push_back(iter.strides(offset)[i]);
+
+  for (int32_t i = 0; i < num_outputs; i++) {
+    uint32_t idx = i;
+    for (int dim = 0; dim < nDim; dim++) {
+      uint32_t remainder = idx % iterShape[dim];
+      idx /= iterShape[dim];
+
+      for (int offset = 0; offset < nOffsets; offset++) {
+        data_offsets[i][offset] += remainder * strides[dim][offset];
+      }
+    }
+  }
+}
+
+static
+bool dispatchIndexSelectKernel(TensorIteratorBase& iter, IntArrayRef index_size, IntArrayRef index_stride) {
+  using namespace mps;
+
+  @autoreleasepool {
+    if (iter.numel() == 0) {
+      return true;
+    }
+
+    const Tensor& inputTensor = iter.tensor(1);
+    Tensor outputTensor = iter.tensor(0);
+
+    id<MTLBuffer> inputBuffer  = getMTLBufferStorage(inputTensor);
+    id<MTLBuffer> outputBuffer = getMTLBufferStorage(outputTensor);
+    MTLResourceOptions options = [inputBuffer resourceOptions];
+    int64_t iter_numel = iter.numel();
+    int64_t num_indices = index_size.size();
+
+    std::string indexFunction;
+    if (!selectIndexFunctionName(inputTensor.scalar_type(), indexFunction)) {
+        TORCH_CHECK(false, "Failed to create indexing library, error: ", inputTensor.scalar_type());
+    }
+
+    NSError* error = nil;
+    MPSStream* mpsStream = getCurrentMPSStream();
+    id<MTLDevice> device = MPSDevice::getInstance()->device();
+
+    id<MTLBuffer> offsets = [device newBufferWithLength: iter_numel * sizeof(simd_uint3)
+                                                options: 0];
+    // Get a list of offsets to index the Input, Output and the Index tensors
+    simd_uint3* data_offsets = (simd_uint3*)[offsets contents];
+    getDataOffsets(iter, data_offsets, iter_numel);
+
+    id<MTLCommandBuffer> commandBuffer = mpsStream->commandBuffer();
+    id<MTLComputeCommandEncoder> computeEncoder = [commandBuffer computeCommandEncoder];
+
+    MTLFunctionConstantValues* constantValues = [MTLFunctionConstantValues new];
+    int64_t storage_offset = inputTensor.storage_offset();
+    [constantValues setConstantValue: &storage_offset type:MTLDataTypeLong atIndex:0];
```

Comment:
Why does it have to be passed as argument to shader rather than via `offset:` argument of `-[MTLComputeCommandEncoder setBuffer:offset:atIndex:]`?
```suggestion
```

Reply:
Fixed in next [PR](https://github.com/kulinseth/pytorch/pull/77/)